tests/python/test_ad_for.py

```diff
@@ -775,3 +775,159 @@ def test_large_loop():
 
     assert loss[None] == 1e7
     assert x.grad[None] == 1e7
+
+
+@test_utils.test(require=ti.extension.adstack)
+def test_multiple_ib():
+    x = ti.field(float, (), needs_grad=True)
+    y = ti.field(float, (), needs_grad=True)
+
+    @ti.kernel
+    def compute_y():
+        for j in range(2):
+            for i in range(3):
+                y[None] += x[None]
+            for i in range(3):
+                y[None] += x[None]
+
+    x[None] = 1.0
+    with ti.Tape(y):
+        compute_y()
+
+    assert y[None] == 12.0
+    assert x.grad[None] == 12.0
+
+
+@test_utils.test(require=ti.extension.adstack)
+def test_multiple_ib_multiple_outermost():
+    x = ti.field(float, (), needs_grad=True)
+    y = ti.field(float, (), needs_grad=True)
+
+    @ti.kernel
+    def compute_y():
+        for j in range(2):
+            for i in range(3):
+                y[None] += x[None]
+            for i in range(3):
+                y[None] += x[None]
+        for j in range(2):
+            for i in range(3):
+                y[None] += x[None]
+            for i in range(3):
+                y[None] += x[None]
+
+    x[None] = 1.0
+    with ti.Tape(y):
+        compute_y()
+
+    assert y[None] == 24.0
+    assert x.grad[None] == 24.0
+
+
+@test_utils.test(require=ti.extension.adstack)
+def test_multiple_ib_multiple_outermost_mixed():
+    x = ti.field(float, (), needs_grad=True)
+    y = ti.field(float, (), needs_grad=True)
+
+    @ti.kernel
+    def compute_y():
+        for j in range(2):
+            for i in range(3):
+                y[None] += x[None]
+            for i in range(3):
+                y[None] += x[None]
+        for j in range(2):
+            for i in range(3):
+                y[None] += x[None]
+            for i in range(3):
+                y[None] += x[None]
+                for ii in range(3):
+                    y[None] += x[None]
+
+    x[None] = 1.0
+    with ti.Tape(y):
+        compute_y()
+
+    assert y[None] == 42.0
+    assert x.grad[None] == 42.0
+
+
+@test_utils.test(require=ti.extension.adstack)
+def test_multiple_ib_mixed():
+    x = ti.field(float, (), needs_grad=True)
+    y = ti.field(float, (), needs_grad=True)
+
+    @ti.kernel
+    def compute_y():
+        for j in range(2):
+            for i in range(3):
+                y[None] += x[None]
+            for i in range(3):
+                y[None] += x[None]
+                for k in range(2):
+                    y[None] += x[None]
+            for i in range(3):
+                y[None] += x[None]
+
+    x[None] = 1.0
+    with ti.Tape(y):
+        compute_y()
+
+    assert y[None] == 30.0
+    assert x.grad[None] == 30.0
+
+
+@test_utils.test(require=ti.extension.adstack)
+def test_multiple_ib_deeper():
```

Comment:
It'd be great to add a test case that the input field isn't a scalar, just to make sure our test coverage is better ;)